src/crimson/thread/ThreadPool.h

```diff
@@ -0,0 +1,95 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+
+#include <atomic>
+#include <condition_variable>
+#include <tuple>
+#include <type_traits>
+#include <boost/lockfree/queue.hpp>
+#include <boost/optional.hpp>
+#include <core/future.hh>
+#include <core/semaphore.hh>
+
+#include "Condition.h"
+
+namespace ceph::thread {
+
+struct WorkItem {
+  virtual ~WorkItem() {}
+  virtual void process() = 0;
+};
+
+template<typename Func, typename T = std::invoke_result_t<Func>>
+struct Task final : WorkItem {
+  Func func;
+  boost::optional<T> result;
+  ceph::thread::Condition on_done;
+  seastar::semaphore& free_slot;
+public:
+  Task(seastar::semaphore& fs, Func&& f)
+    : func(std::move(f)),
+      free_slot(fs)
+  {}
+  void process() override {
+    result = func();
+    on_done.notify();
+  }
+  seastar::future<T> get_future() {
+    return on_done.wait().then([this] {
+        free_slot.signal();
+	auto res = std::move(result.get());
+	delete this;
+	return res;
+      }).then([](T&& result) {
+        return seastar::make_ready_future<T>(std::move(result));
+      });
+  }
+};
+
+/// an engine for scheduling non-seastar tasks from seastar threads
+class ThreadPool {
+  std::atomic<bool> stop = false;
+  std::mutex mutex;
+  std::condition_variable cond;
+  std::vector<std::thread> threads;
+  std::vector<seastar::semaphore> free_slots;
+  // please note, each Task has its own ceph::thread::Condition, which
+  // possesses a fd, so we should keep the number of WorkItem in-flight under a
+  // reasonable limit.
+  static constexpr size_t num_queue_size = 128;
+  using item_queue_t =
+    boost::lockfree::queue<WorkItem*,
+			   boost::lockfree::capacity<num_queue_size>>;
+  item_queue_t pending;
+
+  std::vector<std::thread> create_threads(size_t n);
+  void loop();
+  bool stopping() const {
+    return stop.load(std::memory_order_relaxed);
+  }
+
+  ThreadPool(const ThreadPool&) = delete;
+  ThreadPool& operator=(const ThreadPool&) = delete;
+public:
+  explicit ThreadPool(size_t n);
+  ~ThreadPool();
+  template<typename Func, typename...Args>
+  auto submit(Func&& func, Args&&... args) {
+    auto& free_slot = free_slots[seastar::engine().cpu_id()];
+    auto task = new Task{
```

Comment:
this is allocating Tasks before a slot is available, so we're effectively losing control over the number of outstanding file descriptors that `num_queue_size` was intended to give us

i think we could get away with a single `Condition` per cpu, if we also added a per-cpu sequence number. for example, `submit()` chooses a `task_seq = ++max_seq`, and `get_future()` keeps waiting on the condition until `processed_seq >= task_seq`. does that make sense?

Reply:
the single-`Condition`-per-cpu approach works if multiple fiber in seastar can wait on the same fd, but in seastar, only a single fiber will be waken up. also, we can not use this single task in-flight, as it limits the parallel level of alien thread on this core. i want all cores to share the the full alien thread pool, in other words, a core can use all the threads if other cores are not using any thread in the thread pool.